Lifetimes seem broken

Prompted by https://github.com/rusterlium/rustler/pull/425#discussion_r803431410 I tried to add tests for lifetimes, but I couldn't do it because I got a compile error:

```
    Checking rustler_test v0.1.0 (/home/turion/Nextcloud/e/informatisch/rust/rustler/rustler_tests/native/rustler_test)
error[E0261]: use of undeclared lifetime name `'a`
   --> rustler_tests/native/rustler_test/src/test_codegen.rs:145:10
    |
145 | #[derive(NifStruct)]
    |          -^^^^^^^^
    |          |
    |          undeclared lifetime
    |          lifetime `'a` is missing in item created through this procedural macro
    |
    = note: this error originates in the derive macro `NifStruct` (in Nightly builds, run with -Z macro-backtrace for more info)

For more information about this error, try `rustc --explain E0261`.
```

Whole code:

<details>

```
diff --git rustler_tests/lib/rustler_test.ex rustler_tests/lib/rustler_test.ex
index ec351e2..57e87d7 100644
--- rustler_tests/lib/rustler_test.ex
+++ rustler_tests/lib/rustler_test.ex
@@ -71,6 +71,7 @@ defmodule RustlerTest do
   def tuplestruct_echo(_), do: err()
   def newtype_record_echo(_), do: err()
   def tuplestruct_record_echo(_), do: err()
+  def string_slice_lifetime_echo(), do: err()
   def reserved_keywords_type_echo(_), do: err()
 
   def dirty_io(), do: err()
diff --git rustler_tests/native/rustler_test/src/lib.rs rustler_tests/native/rustler_test/src/lib.rs
index 402637f..808a5ac 100644
--- rustler_tests/native/rustler_test/src/lib.rs
+++ rustler_tests/native/rustler_test/src/lib.rs
@@ -67,6 +67,7 @@ rustler::init!(
         test_codegen::tuplestruct_echo,
         test_codegen::newtype_record_echo,
         test_codegen::tuplestruct_record_echo,
+        test_codegen::string_slice_lifetime,
         test_dirty::dirty_cpu,
         test_dirty::dirty_io,
         test_range::sum_range,
diff --git rustler_tests/native/rustler_test/src/test_codegen.rs rustler_tests/native/rustler_test/src/test_codegen.rs
index 748934e..b0dda3a 100644
--- rustler_tests/native/rustler_test/src/test_codegen.rs
+++ rustler_tests/native/rustler_test/src/test_codegen.rs
@@ -142,6 +142,17 @@ pub fn tuplestruct_record_echo(tuplestruct: TupleStructRecord) -> TupleStructRec
     tuplestruct
 }
 
+#[derive(NifStruct)]
+#[module = "StringSliceLifetime"]
+struct StringSliceLifetime<'a> {
+    message: &'a str,
+}
+
+#[rustler::nif]
+pub fn string_slice_lifetime() -> StringSliceLifetime<'static> {
+    StringSliceLifetime { message: "hi" }
+}
+
 pub mod reserved_keywords {
     use rustler::{NifMap, NifRecord, NifStruct, NifTuple, NifUntaggedEnum};
 
diff --git rustler_tests/test/codegen_test.exs rustler_tests/test/codegen_test.exs
index 2fe5684..855594c 100644
--- rustler_tests/test/codegen_test.exs
+++ rustler_tests/test/codegen_test.exs
@@ -21,6 +21,10 @@ defmodule TupleStructRecord do
   defrecord :tuplestruct, a: 1, b: 2, c: 3
 end
 
+defmodule StringSliceLifetime do
+  defstruct :message
+end
+
 defmodule RustlerTest.CodegenTest do
   use ExUnit.Case, async: true
 
@@ -231,6 +235,11 @@ defmodule RustlerTest.CodegenTest do
                  end
   end
 
+  test "string_slice_lifetime" do
+    value = %StringSliceLifetime{message: "hello"}
+    assert value == string_slice_lifetime_echo()
+  end
+
   test "reserved keywords" do
     assert %{override: 1} == RustlerTest.reserved_keywords_type_echo(%{override: 1})
```

</details>

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,4 +8,5 @@ members = [
   "rustler_tests/native/rustler_test",
   "rustler_tests/native/rustler_bigint_test",
   "rustler_tests/native/deprecated_macros",
+  "rustler_tests/native/rustler_compile_tests",
 ]

--- a/rustler/src/types/binary.rs
+++ b/rustler/src/types/binary.rs
@@ -102,7 +102,7 @@ use std::{
 /// See [module-level doc](index.html) for more information.
 pub struct OwnedBinary(ErlNifBinary);
 
-impl<'a> OwnedBinary {
+impl OwnedBinary {
     pub unsafe fn from_raw(inner: ErlNifBinary) -> OwnedBinary {
         OwnedBinary(inner)
     }

--- a/rustler/src/types/list.rs
+++ b/rustler/src/types/list.rs
@@ -93,7 +93,7 @@ impl<'a> Decoder<'a> for ListIterator<'a> {
 //    }
 //}
 
-impl<'a, T> Encoder for Vec<T>
+impl<T> Encoder for Vec<T>
 where
     T: Encoder,
 {
@@ -113,7 +113,7 @@ where
     }
 }
 
-impl<'a, T> Encoder for [T]
+impl<T> Encoder for [T]
 where
     T: Encoder,
 {

--- a/rustler/src/wrapper/binary.rs
+++ b/rustler/src/wrapper/binary.rs
@@ -1,5 +1,5 @@
 use crate::{wrapper::size_t, Env, Term};
-pub(in crate) use rustler_sys::ErlNifBinary;
+pub(crate) use rustler_sys::ErlNifBinary;
 use std::mem::MaybeUninit;
 
 pub unsafe fn alloc(size: size_t) -> Option<ErlNifBinary> {

--- a/rustler_bigint/src/big_int.rs
+++ b/rustler_bigint/src/big_int.rs
@@ -59,7 +59,7 @@ rustler::atoms! {
 /// }
 /// ```
 ///
-#[derive(Debug, PartialEq, PartialOrd)]
+#[derive(Debug, PartialEq, PartialOrd, Eq)]
 pub struct BigInt(num_bigint::BigInt);
 
 impl std::convert::From<num_bigint::BigInt> for BigInt {
@@ -89,7 +89,7 @@ impl std::ops::DerefMut for BigInt {
 }
 
 fn decode_big_integer(input: &[u8]) -> NifResult<num_bigint::BigInt> {
-    if Some(&EXTERNAL_TERM_FORMAT_VERSION) != input.get(0) {
+    if Some(&EXTERNAL_TERM_FORMAT_VERSION) != input.first() {
         return Err(Error::BadArg);
     }
 

--- a/rustler_codegen/src/context.rs
+++ b/rustler_codegen/src/context.rs
@@ -16,7 +16,7 @@ use super::RustlerAttr;
 pub(crate) struct Context<'a> {
     pub attrs: Vec<RustlerAttr>,
     pub ident: &'a proc_macro2::Ident,
-    pub ident_with_lifetime: proc_macro2::TokenStream,
+    pub generics: &'a syn::Generics,
     pub variants: Option<Vec<&'a Variant>>,
     pub struct_fields: Option<Vec<&'a Field>>,
     pub is_tuple_struct: bool,
@@ -38,19 +38,6 @@ impl<'a> Context<'a> {
             attrs.push(RustlerAttr::Decode);
         }
 
-        let has_lifetime = match ast.generics.lifetimes().count() {
-            0 => false,
-            1 => true,
-            _ => panic!("Struct can only have one lifetime argument"),
-        };
-
-        let ident = &ast.ident;
-        let ident_with_lifetime = if has_lifetime {
-            quote! { #ident <'a> }
-        } else {
-            quote! { #ident }
-        };
-
         let variants = match ast.data {
             Data::Enum(ref data_enum) => Some(data_enum.variants.iter().collect()),
             _ => None,
@@ -71,8 +58,8 @@ impl<'a> Context<'a> {
 
         Self {
             attrs,
-            ident,
-            ident_with_lifetime,
+            ident: &ast.ident,
+            generics: &ast.generics,
             variants,
             struct_fields,
             is_tuple_struct,

--- a/rustler_codegen/src/encode_decode_templates.rs
+++ b/rustler_codegen/src/encode_decode_templates.rs
@@ -1,0 +1,81 @@
+use proc_macro2::{Span, TokenStream};
+use quote::quote;
+
+use super::context::Context;
+
+pub(crate) fn decoder(ctx: &Context, inner: TokenStream) -> TokenStream {
+    let ident = ctx.ident;
+    let generics = ctx.generics;
+    let (_impl_generics, ty_generics, _where_clause) = generics.split_for_impl();
+
+    // The Decoder uses a special lifetime '__rustler_decode_lifetime. We need to ensure that all
+    // other lifetimes are bound to this lifetime: As we decode from a term (which has a lifetime),
+    // references to that term may not outlive the term itself.
+    let lifetimes: Vec<_> = generics
+        .params
+        .iter()
+        .filter_map(|g| match g {
+            syn::GenericParam::Lifetime(l) => Some(l.lifetime.clone()),
+            _ => None,
+        })
+        .collect();
+
+    let mut impl_generics = generics.clone();
+    let decode_lifetime = syn::Lifetime::new("'__rustler_decode_lifetime", Span::call_site());
+    let lifetime_def = syn::LifetimeDef::new(decode_lifetime.clone());
+    impl_generics
+        .params
+        .push(syn::GenericParam::Lifetime(lifetime_def));
+
+    if !lifetimes.is_empty() {
+        let where_clause = impl_generics.make_where_clause();
+
+        for lifetime in lifetimes {
+            let mut puncated = syn::punctuated::Punctuated::new();
+            puncated.push(lifetime.clone());
+            let predicate = syn::PredicateLifetime {
+                lifetime: decode_lifetime.clone(),
+                colon_token: syn::token::Colon {
+                    spans: [Span::call_site()],
+                },
+                bounds: puncated,
+            };
+            where_clause.predicates.push(predicate.into());
+
+            let mut puncated = syn::punctuated::Punctuated::new();
+            puncated.push(decode_lifetime.clone());
+            let predicate = syn::PredicateLifetime {
+                lifetime: lifetime.clone(),
+                colon_token: syn::token::Colon {
+                    spans: [Span::call_site()],
+                },
+                bounds: puncated,
+            };
+            where_clause.predicates.push(predicate.into());
+        }
+    }
+
+    let (impl_generics, _, where_clause) = impl_generics.split_for_impl();
+
+    quote! {
+        impl #impl_generics ::rustler::Decoder<'__rustler_decode_lifetime> for #ident #ty_generics #where_clause {
+            fn decode(term: ::rustler::Term<'__rustler_decode_lifetime>) -> ::rustler::NifResult<Self> {
+                #inner
+            }
+        }
+    }
+}
+
+pub(crate) fn encoder(ctx: &Context, inner: TokenStream) -> TokenStream {
+    let ident = ctx.ident;
+    let generics = ctx.generics;
+    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
+
+    quote! {
+        impl #impl_generics ::rustler::Encoder for #ident #ty_generics #where_clause {
+            fn encode<'__rustler__encode_lifetime>(&self, env: ::rustler::Env<'__rustler__encode_lifetime>) -> ::rustler::Term<'__rustler__encode_lifetime> {
+                #inner
+            }
+        }
+    }
+}

--- a/rustler_codegen/src/ex_struct.rs
+++ b/rustler_codegen/src/ex_struct.rs
@@ -65,7 +65,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput, add_exception: bool) -> Toke
 }
 
 fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
     let struct_name = ctx.ident;
     let struct_name_str = struct_name.to_string();
 
@@ -94,42 +93,39 @@ fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> T
         })
         .unzip();
 
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #struct_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
-                use #atoms_module_name::*;
-                use ::rustler::Encoder;
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
+            use ::rustler::Encoder;
 
-                fn try_decode_field<'a, T>(
-                    term: rustler::Term<'a>,
-                    field: rustler::Atom,
-                    ) -> ::rustler::NifResult<T>
-                    where
-                        T: rustler::Decoder<'a>,
-                    {
-                        use rustler::Encoder;
-                        match ::rustler::Decoder::decode(term.map_get(&field)?) {
-                            Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
-                                            "Could not decode field :{:?} on %{}{{}}",
-                                            field, #struct_name_str
-                            )))),
-                            Ok(value) => Ok(value),
-                        }
-                    };
-
-                let module: ::rustler::types::atom::Atom = term.map_get(atom_struct())?.decode()?;
-                if module != atom_module() {
-                    return Err(::rustler::Error::RaiseAtom("invalid_struct"));
+            fn try_decode_field<'a, T>(
+                term: rustler::Term<'a>,
+                field: rustler::Atom,
+                ) -> ::rustler::NifResult<T>
+                where
+                    T: rustler::Decoder<'a>,
+                {
+                    use rustler::Encoder;
+                    match ::rustler::Decoder::decode(term.map_get(&field)?) {
+                        Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
+                                        "Could not decode field :{:?} on %{}{{}}",
+                                        field, #struct_name_str
+                        )))),
+                        Ok(value) => Ok(value),
+                    }
                 }
 
-                #(#assignments);*
-
-                Ok(#struct_name { #(#field_defs),* })
+            let module: ::rustler::types::atom::Atom = term.map_get(atom_struct())?.decode()?;
+            if module != atom_module() {
+                return Err(::rustler::Error::RaiseAtom("invalid_struct"));
             }
-        }
-    };
 
-    gen
+            #(#assignments);*
+
+            Ok(#struct_name { #(#field_defs),* })
+        },
+    )
 }
 
 fn gen_encoder(
@@ -138,8 +134,6 @@ fn gen_encoder(
     atoms_module_name: &Ident,
     add_exception: bool,
 ) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
-
     let field_defs: Vec<TokenStream> = fields
         .iter()
         .map(|field| {
@@ -159,20 +153,17 @@ fn gen_encoder(
         quote! {}
     };
 
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #struct_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                use #atoms_module_name::*;
-                let mut map = ::rustler::types::map::map_new(env);
-                map = map.map_put(atom_struct(), atom_module()).unwrap();
-                #exception_field
-                #(#field_defs)*
-                map
-            }
-        }
-    };
-
-    gen
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
+            let mut map = ::rustler::types::map::map_new(env);
+            map = map.map_put(atom_struct(), atom_module()).unwrap();
+            #exception_field
+            #(#field_defs)*
+            map
+        },
+    )
 }
 
 fn get_module(ctx: &Context, add_exception: bool) -> String {

--- a/rustler_codegen/src/lib.rs
+++ b/rustler_codegen/src/lib.rs
@@ -3,6 +3,7 @@
 use proc_macro::TokenStream;
 
 mod context;
+mod encode_decode_templates;
 mod ex_struct;
 mod init;
 mod map;

--- a/rustler_codegen/src/map.rs
+++ b/rustler_codegen/src/map.rs
@@ -49,7 +49,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 }
 
 fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
     let struct_name = ctx.ident;
 
     let idents: Vec<_> = fields
@@ -76,41 +75,36 @@ fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> T
         })
         .unzip();
 
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #struct_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                fn try_decode_field<'a, T>(
-                    term: rustler::Term<'a>,
-                    field: rustler::Atom,
-                    ) -> ::rustler::NifResult<T>
-                    where
-                        T: rustler::Decoder<'a>,
-                    {
-                        use rustler::Encoder;
-                        match ::rustler::Decoder::decode(term.map_get(&field)?) {
-                            Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
-                                            "Could not decode field :{:?} on %{{}}",
-                                            field
-                            )))),
-                            Ok(value) => Ok(value),
-                        }
-                    };
+            fn try_decode_field<'a, T>(
+                term: rustler::Term<'a>,
+                field: rustler::Atom,
+                ) -> ::rustler::NifResult<T>
+                where
+                    T: rustler::Decoder<'a>,
+                {
+                    use rustler::Encoder;
+                    match ::rustler::Decoder::decode(term.map_get(&field)?) {
+                        Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
+                                        "Could not decode field :{:?} on %{{}}",
+                                        field
+                        )))),
+                        Ok(value) => Ok(value),
+                    }
+                };
 
-                #(#assignments);*
+            #(#assignments);*
 
-                Ok(#struct_name { #(#field_defs),* })
-            }
-        }
-    };
-
-    gen
+            Ok(#struct_name { #(#field_defs),* })
+        },
+    )
 }
 
 fn gen_encoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
-
     let field_defs: Vec<TokenStream> = fields
         .iter()
         .map(|field| {
@@ -123,17 +117,14 @@ fn gen_encoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> T
         })
         .collect();
 
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #struct_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                let mut map = ::rustler::types::map::map_new(env);
-                #(#field_defs)*
-                map
-            }
-        }
-    };
-
-    gen
+            let mut map = ::rustler::types::map::map_new(env);
+            #(#field_defs)*
+            map
+        },
+    )
 }

--- a/rustler_codegen/src/record.rs
+++ b/rustler_codegen/src/record.rs
@@ -49,7 +49,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 }
 
 fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
     let struct_name = ctx.ident;
 
     // Make a decoder for each of the fields in the struct.
@@ -97,49 +96,45 @@ fn gen_decoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> T
             Ok(#struct_name { #(#field_defs),* })
         }
     };
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #struct_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
-                use #atoms_module_name::*;
 
-                let terms = match ::rustler::types::tuple::get_tuple(term) {
-                    Err(_) => return Err(::rustler::Error::RaiseTerm(
-                            Box::new(format!("Invalid Record structure for {}", #struct_name_str)))),
-                    Ok(value) => value,
-                };
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                if terms.len() != #field_num + 1 {
-                    return Err(::rustler::Error::RaiseAtom("invalid_record"));
-                }
+            let terms = match ::rustler::types::tuple::get_tuple(term) {
+                Err(_) => return Err(::rustler::Error::RaiseTerm(
+                        Box::new(format!("Invalid Record structure for {}", #struct_name_str)))),
+                Ok(value) => value,
+            };
 
-                let tag : ::rustler::types::atom::Atom = terms[0].decode()?;
-
-                if tag != atom_tag() {
-                    return Err(::rustler::Error::RaiseAtom("invalid_record"));
-                }
-
-                fn try_decode_index<'a, T>(terms: &[::rustler::Term<'a>], pos_in_struct: &str, index: usize) -> ::rustler::NifResult<T>
-                    where
-                        T: rustler::Decoder<'a>,
-                {
-                    match ::rustler::Decoder::decode(terms[index]) {
-                        Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(
-                                    format!("Could not decode field {} on Record {}", pos_in_struct, #struct_name_str)))),
-                        Ok(value) => Ok(value)
-                    }
-                }
-
-                #construct
+            if terms.len() != #field_num + 1 {
+                return Err(::rustler::Error::RaiseAtom("invalid_record"));
             }
-        }
-    };
 
-    gen
+            let tag : ::rustler::types::atom::Atom = terms[0].decode()?;
+
+            if tag != atom_tag() {
+                return Err(::rustler::Error::RaiseAtom("invalid_record"));
+            }
+
+            fn try_decode_index<'a, T>(terms: &[::rustler::Term<'a>], pos_in_struct: &str, index: usize) -> ::rustler::NifResult<T>
+                where
+                    T: rustler::Decoder<'a>,
+            {
+                match ::rustler::Decoder::decode(terms[index]) {
+                    Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(
+                                format!("Could not decode field {} on Record {}", pos_in_struct, #struct_name_str)))),
+                    Ok(value) => Ok(value)
+                }
+            }
+
+            #construct
+        },
+    )
 }
 
 fn gen_encoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
-
     // Make a field encoder expression for each of the items in the struct.
     let field_encoders: Vec<TokenStream> = fields
         .iter()
@@ -163,20 +158,16 @@ fn gen_encoder(ctx: &Context, fields: &[&Field], atoms_module_name: &Ident) -> T
         [#tag_encoder, #(#field_encoders),*]
     };
 
-    // The implementation itself
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #struct_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                use ::rustler::Encoder;
-                let arr = #field_list_ast;
-                ::rustler::types::tuple::make_tuple(env, &arr)
-            }
-        }
-    };
-
-    gen
+            use ::rustler::Encoder;
+            let arr = #field_list_ast;
+            ::rustler::types::tuple::make_tuple(env, &arr)
+        },
+    )
 }
 
 fn get_tag(ctx: &Context) -> String {

--- a/rustler_codegen/src/tagged_enum.rs
+++ b/rustler_codegen/src/tagged_enum.rs
@@ -83,7 +83,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 }
 
 fn gen_decoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) -> TokenStream {
-    let enum_type = &ctx.ident_with_lifetime;
     let enum_name = ctx.ident;
     let unit_decoders: Vec<TokenStream> = variants
         .iter()
@@ -117,48 +116,44 @@ fn gen_decoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) 
         })
         .collect();
 
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #enum_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                fn try_decode_field<'a, T>(
-                    term: ::rustler::Term<'a>,
-                    field: ::rustler::Atom,
-                    ) -> ::rustler::NifResult<T>
-                    where
-                        T: ::rustler::Decoder<'a>,
-                {
-                    use ::rustler::Encoder;
-                    match ::rustler::Decoder::decode(term.map_get(&field)?) {
-                        Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
-                                        "Could not decode field :{:?} on %{{}}",
-                                        field
-                        )))),
-                        Ok(value) => Ok(value),
-                    }
+            fn try_decode_field<'a, T>(
+                term: ::rustler::Term<'a>,
+                field: ::rustler::Atom,
+                ) -> ::rustler::NifResult<T>
+                where
+                    T: ::rustler::Decoder<'a>,
+            {
+                use ::rustler::Encoder;
+                match ::rustler::Decoder::decode(term.map_get(&field)?) {
+                    Err(_) => Err(::rustler::Error::RaiseTerm(Box::new(format!(
+                                    "Could not decode field :{:?} on %{{}}",
+                                    field
+                    )))),
+                    Ok(value) => Ok(value),
                 }
-
-                if let Ok(unit) = ::rustler::types::atom::Atom::from_term(term) {
-                    #(#unit_decoders)*
-                } else if let Ok(tuple) = ::rustler::types::tuple::get_tuple(term) {
-                    let name = tuple
-                                .get(0)
-                                .and_then(|&first| ::rustler::types::atom::Atom::from_term(first).ok())
-                                .ok_or(::rustler::Error::RaiseAtom("invalid_variant"))?;
-                    #(#named_unnamed_decoders)*
-                }
-
-                Err(::rustler::Error::RaiseAtom("invalid_variant"))
             }
-        }
-    };
 
-    gen
+            if let Ok(unit) = ::rustler::types::atom::Atom::from_term(term) {
+                #(#unit_decoders)*
+            } else if let Ok(tuple) = ::rustler::types::tuple::get_tuple(term) {
+                let name = tuple
+                            .get(0)
+                            .and_then(|&first| ::rustler::types::atom::Atom::from_term(first).ok())
+                            .ok_or(::rustler::Error::RaiseAtom("invalid_variant"))?;
+                #(#named_unnamed_decoders)*
+            }
+
+            Err(::rustler::Error::RaiseAtom("invalid_variant"))
+        },
+    )
 }
 
 fn gen_encoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) -> TokenStream {
-    let enum_type = &ctx.ident_with_lifetime;
     let enum_name = ctx.ident;
 
     let variant_defs: Vec<TokenStream> = variants
@@ -179,19 +174,16 @@ fn gen_encoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) 
         })
         .collect();
 
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #enum_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                match self {
-                    #(#variant_defs)*
-                }
+            match self {
+                #(#variant_defs)*
             }
-        }
-    };
-
-    gen
+        },
+    )
 }
 
 fn gen_unit_decoder(enum_name: &Ident, variant_ident: &Ident, atom_fn: Ident) -> TokenStream {

--- a/rustler_codegen/src/tuple.rs
+++ b/rustler_codegen/src/tuple.rs
@@ -34,7 +34,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 }
 
 fn gen_decoder(ctx: &Context, fields: &[&Field]) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
     let struct_name = ctx.ident;
     let struct_name_str = struct_name.to_string();
 
@@ -81,9 +80,10 @@ fn gen_decoder(ctx: &Context, fields: &[&Field]) -> TokenStream {
             Ok(#struct_name { #(#field_defs),* })
         }
     };
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #struct_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
+
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
                 let terms = ::rustler::types::tuple::get_tuple(term)?;
                 if terms.len() != #field_num {
                     return Err(::rustler::Error::BadArg);
@@ -99,17 +99,13 @@ fn gen_decoder(ctx: &Context, fields: &[&Field]) -> TokenStream {
                         Ok(value) => Ok(value)
                     }
                 }
-                #construct
-            }
-        }
-    };
 
-    gen
+                #construct
+        },
+    )
 }
 
 fn gen_encoder(ctx: &Context, fields: &[&Field]) -> TokenStream {
-    let struct_type = &ctx.ident_with_lifetime;
-
     // Make a field encoder expression for each of the items in the struct.
     let field_encoders: Vec<TokenStream> = fields
         .iter()
@@ -130,16 +126,12 @@ fn gen_encoder(ctx: &Context, fields: &[&Field]) -> TokenStream {
         [#(#field_encoders),*]
     };
 
-    // The implementation itself
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #struct_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                use ::rustler::Encoder;
-                let arr = #field_list_ast;
-                ::rustler::types::tuple::make_tuple(env, &arr)
-            }
-        }
-    };
-
-    gen
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            use ::rustler::Encoder;
+            let arr = #field_list_ast;
+            ::rustler::types::tuple::make_tuple(env, &arr)
+        },
+    )
 }

--- a/rustler_codegen/src/unit_enum.rs
+++ b/rustler_codegen/src/unit_enum.rs
@@ -67,7 +67,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 }
 
 fn gen_decoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) -> TokenStream {
-    let enum_type = &ctx.ident_with_lifetime;
     let enum_name = ctx.ident;
 
     let variant_defs: Vec<TokenStream> = variants
@@ -85,25 +84,21 @@ fn gen_decoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) 
         })
         .collect();
 
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #enum_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                let value = ::rustler::types::atom::Atom::from_term(term)?;
+            let value = ::rustler::types::atom::Atom::from_term(term)?;
 
-                #(#variant_defs)*
+            #(#variant_defs)*
 
-                Err(::rustler::Error::RaiseAtom("invalid_variant"))
-            }
-        }
-    };
-
-    gen
+            Err(::rustler::Error::RaiseAtom("invalid_variant"))
+        },
+    )
 }
 
 fn gen_encoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) -> TokenStream {
-    let enum_type = &ctx.ident_with_lifetime;
     let enum_name = ctx.ident;
 
     let variant_defs: Vec<TokenStream> = variants
@@ -119,17 +114,14 @@ fn gen_encoder(ctx: &Context, variants: &[&Variant], atoms_module_name: &Ident) 
         })
         .collect();
 
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #enum_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                use #atoms_module_name::*;
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            use #atoms_module_name::*;
 
-                match *self {
-                    #(#variant_defs)*
-                }
+            match *self {
+                #(#variant_defs)*
             }
-        }
-    };
-
-    gen
+        },
+    )
 }

--- a/rustler_codegen/src/untagged_enum.rs
+++ b/rustler_codegen/src/untagged_enum.rs
@@ -48,7 +48,6 @@ pub fn transcoder_decorator(ast: &syn::DeriveInput) -> TokenStream {
 }
 
 fn gen_decoder(ctx: &Context, variants: &[&Variant]) -> TokenStream {
-    let enum_type = &ctx.ident_with_lifetime;
     let enum_name = ctx.ident;
 
     let variant_defs: Vec<_> = variants
@@ -65,21 +64,17 @@ fn gen_decoder(ctx: &Context, variants: &[&Variant]) -> TokenStream {
         })
         .collect();
 
-    let gen = quote! {
-        impl<'a> ::rustler::Decoder<'a> for #enum_type {
-            fn decode(term: ::rustler::Term<'a>) -> ::rustler::NifResult<Self> {
-                #(#variant_defs)*
+    super::encode_decode_templates::decoder(
+        ctx,
+        quote! {
+            #(#variant_defs)*
 
-                Err(::rustler::Error::RaiseAtom("invalid_variant"))
-            }
-        }
-    };
-
-    gen
+            Err(::rustler::Error::RaiseAtom("invalid_variant"))
+        },
+    )
 }
 
 fn gen_encoder(ctx: &Context, variants: &[&Variant]) -> TokenStream {
-    let enum_type = &ctx.ident_with_lifetime;
     let enum_name = ctx.ident;
 
     let variant_defs: Vec<_> = variants
@@ -93,15 +88,12 @@ fn gen_encoder(ctx: &Context, variants: &[&Variant]) -> TokenStream {
         })
         .collect();
 
-    let gen = quote! {
-        impl<'b> ::rustler::Encoder for #enum_type {
-            fn encode<'a>(&self, env: ::rustler::Env<'a>) -> ::rustler::Term<'a> {
-                match *self {
-                    #(#variant_defs)*
-                }
+    super::encode_decode_templates::encoder(
+        ctx,
+        quote! {
+            match *self {
+                #(#variant_defs)*
             }
-        }
-    };
-
-    gen
+        },
+    )
 }

--- a/rustler_sys/build.rs
+++ b/rustler_sys/build.rs
@@ -22,7 +22,7 @@ trait ApiBuilder {
     fn dummy(&mut self, name: &str);
 }
 
-#[derive(PartialEq, Clone, Copy)]
+#[derive(PartialEq, Eq, Clone, Copy)]
 pub enum OsFamily {
     Unix,
     Win,

--- a/rustler_tests/native/rustler_compile_tests/Cargo.toml
+++ b/rustler_tests/native/rustler_compile_tests/Cargo.toml
@@ -1,0 +1,14 @@
+[package]
+name = "rustler_compile_tests"
+version = "0.1.0"
+authors = []
+edition = "2018"
+
+[lib]
+name = "rustler_test"
+path = "src/lib.rs"
+crate-type = ["cdylib"]
+
+[dependencies]
+lazy_static = "1.4"
+rustler = { path = "../../../rustler" }

--- a/rustler_tests/native/rustler_compile_tests/src/lib.rs
+++ b/rustler_tests/native/rustler_compile_tests/src/lib.rs
@@ -1,0 +1,41 @@
+pub mod lifetimes {
+    use rustler::{Binary, NifMap, NifRecord, NifStruct, NifTuple};
+
+    #[derive(NifMap)]
+    pub struct GenericMap<'a, 'b> {
+        pub i: Binary<'a>,
+        pub j: Binary<'b>,
+    }
+
+    #[derive(NifStruct)]
+    #[module = "GenericStruct"]
+    pub struct GenericStruct<'a, 'b> {
+        pub i: Binary<'a>,
+        pub j: Binary<'b>,
+    }
+
+    #[derive(NifRecord)]
+    #[tag = "generic_record"]
+    pub struct GenericRecord<'a, 'b> {
+        pub i: Binary<'a>,
+        pub j: Binary<'b>,
+    }
+
+    #[derive(NifTuple)]
+    pub struct GenericTuple<'a, 'b>(Binary<'a>, Binary<'b>);
+
+    #[derive(NifMap)]
+    pub struct WhereClause<'a, 'b>
+    where
+        'a: 'b,
+    {
+        pub i: Binary<'a>,
+        pub j: Binary<'b>,
+    }
+
+    #[derive(NifMap)]
+    pub struct LifetimeBounded<'a, 'b: 'a> {
+        pub i: Binary<'a>,
+        pub j: Binary<'b>,
+    }
+}

